### ISSUE TITLE
Improve build chain

### ISSUE THREAD

**Issue by paxter** [porla/porla#326]
It's nearly impossible to compile Porla with latest updates of dependencies on up-to-date systems (i.e. a lot of trouble with latest Lua 5.4.x). In the current state, it works only on very specific systems/setups.

A better build chain with specific documents, required versions of dependencies etc. would be nice. In the current state, it's more a "works on my local system" build chain. Should be more generous for cross-platform support. Additional BSD support would be nice too.

**Comment by vktr** [comment]
I agree somewhat, but using the devcontainer will set you up with a development environment that Works :tm:

**Comment by txtsd** [comment]
I'm trying to get it to build on Arch Linux with system packages. Stay tuned.

**Comment by txtsd** [comment]
@paxter If you're a packager, you could probably gain some insight from here:
https://aur.archlinux.org/packages/porla

**Comment by paxter** [comment]
Nice one @txtsd! I like Arch very much. Unfortunately, my current target machine runs RHEL. Nevertheless, your created arch package will be very helpful on my next compile try. :)

I'm using Arch on another machine. Maybe I will use these and switch to your package. ;)

**Comment by Aubermean** [comment]
Hi @vktr, nice work, trying to have a play around...

Has the docker build environment changed? Is the readme outdated? I have tried within the devcontainer in vscode too, but the build fails. Is `git submodule update --init --recursive` a missing step?

I just want to get a docker image built with a few changes to master branch. Thanks!

**Comment by vktr** [comment]
Hi @Aubermean! Yeah, doing that should yield a workable devcontainer in VS Code. If you're running Linux, I recently added a devenv.sh environment as well. It's not very well tested but I did do the recent Lua metrics integration using that!

**Comment by Aubermean** [comment]
I fell into the rabit hole a bit with devcontainers, a vscode feature I have not used before. For others, less familiar with devcontainers or building in general, here are steps to build a docker container fast after making changes (which will work on any OS).

If you're starting from scratch:

```bash
git clone --recursive https://github.com/porla/porla
cd porla
docker build -t porla:custom
```

If you've already cloned:

```bash
cd porla
git submodule update --init --recursive
docker build -t porla:custom
```

Obviously it can take a while, especially setup-env.sh... would be nice to have more build stages or a prebuilt image to speed things up, since it looks like setup env rarely changes. This helps to rebuild the docker image quickly when making small changes.

Thanks again @vktr, really neat project.

**Comment by vktr** [comment]
We did have prebuilt dependency image before, but it hasn't been used for the devcontainer stuff. A speedy build both locally and on GH Actions would be really nice 😄

**Comment by Aubermean** [comment]
Honestly its more of a skill issue because I don't know how to use devcontainers and I am more familiar with GitLab CI/CD... But I guess my point was that it can help to optimize the multi stage build in the dockerfile, so that code changes don't invalidate the cache and require `setup-env.sh` to be run again... something along the lines of:

```dockerfile
FROM ghcr.io/porla/alpine:3.19.1 AS deps

WORKDIR /src
COPY scripts/setup-env.sh .
RUN ./setup-env.sh

FROM deps AS build-env

WORKDIR /src
COPY . .
RUN cmake -S . -B build -G Ninja -DCMAKE_BUILD_TYPE=Release -DCMAKE_EXE_LINKER_FLAGS="-static -Os" -DBUILD_SHARED_LIBS=OFF -DLINK_WITH_STATIC_LIBRARIES=ON -DOPENSSL_USE_STATIC_LIBS=TRUE -DURIPARSER_SHARED_LIBS=OFF
RUN cmake --build build

FROM ghcr.io/porla/alpine:3.19.1 AS runtime

ENV PORLA_HTTP_HOST=0.0.0.0
EXPOSE 1337

RUN apk --no-cache add curl bash
WORKDIR /
COPY --from=build-env /src/build/porla /usr/bin/porla
ENTRYPOINT [ "/usr/bin/porla" ]
```

Of course it would be even handier to have the build image published, since it doesn't change.

Thanks again for your work, it really is a cool project and I can see it surpassing rutorrent/qbitorrent one day with enough work and exposure. I'm surprised I didn't come across it sooner, I found you via autobrr. Looking forward to making some plugins (Lua <3). I don't care too much about the UI, I guess Flood could be used too in theory...

**Comment by vktr** [comment]
That's what we had a while back - a published build deps image, based on the `Dockerfile.build-env` file, but it hasn't been updated since we removed vcpkg in favor of just building the dependencies ourselves 😄 

Always happy to hear from new users! Feel free to hop into our Discord when you're ready to write some plugins, there's a few of us doing neat things - and also check out https://github.com/porla-plugins/ for some inspiration.

**Comment by vktr** [comment]
I just want to mention this here as well - I've updated the Dockerfile to build each dependency in a separate layer to better utilize the Docker build cache. I've seen very nice improvements.

**Comment by Aubermean** [comment]
Yep - I did the same, forgot to share my working dockerfile.

another optimization I found was setting cpu threads to number of cores for the build. Took. My build time from above an hour to 8 times less!
